src/Microsoft.AspNetCore.JsonPatch/Internal/DictionaryAdapterOfTU.cs

```diff
@@ -18,10 +18,11 @@ public class ExpandoObjectAdapter : IAdapter
         {
             var contract = (JsonDictionaryContract)contractResolver.ResolveContract(target.GetType());
             var key = contract.DictionaryKeyResolver(segment);
-            var dictionary = (IDictionary<string, object>)target;
+            var dictionary = (IDictionary<TKey, TValue>)target;
 
             // As per JsonPatch spec, if a key already exists, adding should replace the existing value
-            dictionary[key] = ConvertValue(dictionary, key, value);
+            var convertedKey = (TKey)ConversionResultProvider.ConvertTo(key, typeof(TKey)).ConvertedInstance;
```

Comment:
What if this conversion fails? Casting a null value to `int` for instance will throw an exception